File: engine/julia/continuation_geometry_pattern_engine.jl
# Module BB5 (Julia): Bilateral & Continuation Geometric Pattern Engine
# Quantitative evaluation of triangle compressions, flags, and continuation projections.

module ContinuationGeometryPatternEngine

export evaluate_triangle, evaluate_flag

function evaluate_triangle(upper_slope::Float64, lower_slope::Float64, base_height::Float64, breakout_px::Float64, spot::Float64)
    if abs(upper_slope) < 0.05 && lower_slope > 0.05
        type_str = "ASCENDING_TRIANGLE"
        target = breakout_px + base_height
        breakout = spot > breakout_px
    elseif abs(lower_slope) < 0.05 && upper_slope < -0.05
        type_str = "DESCENDING_TRIANGLE"
        target = breakout_px - base_height
        breakout = spot < breakout_px
    else
        type_str = "SYMMETRICAL_TRIANGLE"
        target = spot > breakout_px ? (breakout_px + base_height) : (breakout_px - base_height)
        breakout = abs(spot - breakout_px) > 0.5
    end

    return (pattern_type = type_str, target = target, is_breakout = breakout)
end

function evaluate_flag(flag_start::Float64, flag_peak::Float64, breakout_px::Float64, spot::Float64, is_bull::Bool)
    height = abs(flag_peak - flag_start)
    target = is_bull ? (breakout_px + height) : (breakout_px - height)
    breakout = is_bull ? (spot > breakout_px) : (spot < breakout_px)
    return (height = height, target = target, is_breakout = breakout)
end

end

File: engine/rust/src/continuation_geometry_pattern_engine.rs
//! Module BB4 (Rust): Bilateral & Continuation Geometric Pattern Engine.
//! SIMD-optimized evaluation of triangle geometries, flagpoles, and measured move breakouts.

#[repr(C, align(64))]
#[derive(Debug, Clone, Copy)]
pub struct ContinuationGeometryState {
    pub breakout_price: f64,
    pub pattern_dimension_height: f64,
    pub measured_price_target: f64,
    pub current_spot_price: f64,
    pub pattern_geometry_id: u32,
    pub is_breakout_confirmed: u32,
    pub _padding: [u8; 24],
}

impl ContinuationGeometryState {
    pub fn evaluate(geom_id: u32, breakout_px: f64, dim_height: f64, spot: f64, is_bull: u32) -> Self {
        let target = if is_bull == 1 { breakout_px + dim_height } else { breakout_px - dim_height };
        let breakout = if is_bull == 1 {
            if spot > breakout_px { 1 } else { 0 }
        } else {
            if spot < breakout_px { 1 } else { 0 }
        };

        Self {
            breakout_price: breakout_px,
            pattern_dimension_height: dim_height,
            measured_price_target: target,
            current_spot_price: spot,
            pattern_geometry_id: geom_id,
            is_breakout_confirmed: breakout,
            _padding: [0u8; 24],
        }
    }
}
